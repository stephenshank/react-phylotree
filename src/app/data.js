const yokoyama = "(((EELA:0.150276,CONGERA:0.213019):0.230956,(EELB:0.263487,CONGERB:0.202633):0.246917):0.094785,((CAVEFISH:0.451027,(GOLDFISH:0.340495,ZEBRAFISH:0.390163):0.220565):0.067778,((((((NSAM:0.008113,NARG:0.014065):0.052991,SPUN:0.061003,(SMIC:0.027806,SDIA:0.015298,SXAN:0.046873):0.046977):0.009822,(NAUR:0.081298,(SSPI:0.023876,STIE:0.013652):0.058179):0.091775):0.073346,(MVIO:0.012271,MBER:0.039798):0.178835):0.147992,((BFNKILLIFISH:0.317455,(ONIL:0.029217,XCAU:0.084388):0.201166):0.055908,THORNYHEAD:0.252481):0.061905):0.157214,LAMPFISH:0.717196,((SCABBARDA:0.189684,SCABBARDB:0.362015):0.282263,((VIPERFISH:0.318217,BLACKDRAGON:0.109912):0.123642,LOOSEJAW:0.397100):0.287152):0.140663):0.206729):0.222485,(COELACANTH:0.558103,((CLAWEDFROG:0.441842,SALAMANDER:0.299607):0.135307,((CHAMELEON:0.771665,((PIGEON:0.150909,CHICKEN:0.172733):0.082163,ZEBRAFINCH:0.099172):0.272338):0.014055,((BOVINE:0.167569,DOLPHIN:0.157450):0.104783,ELEPHANT:0.166557):0.367205):0.050892):0.114731):0.295021)",
  ksr2 = "((((((HUM,PAN)Node6,GOR)Node5,PON)Node4,GIB)Node3,(MAC,BAB)Node12)Node2,MAR,BUS);",
  h3 = "(KF790010:0.001247,CY147300:0.000623,((CY168239:0.001872,(CY171039:0.001874,KF790389:0.000623)Node7:0.000624)Node5:1e-10,((CY141224:0.001247,(CY147305:1e-10,KF790037:0.000623)Node13:0.000623)Node11:1e-10,(((KF790403:0.000623,KF761507:0.001876)Node18:1e-10,KC526205:0.000623)Node17:1e-10,((CY169303:1e-10,(KF789585:0.000623,CY141202:0.000623)Node25:0.000623)Node23:0.000623,(((KJ667974:0.002501,KJ938675:0.001873)Node30:0.000623,CY134876:1e-10)Node29:0.000623,((KF790414:0.001248,(KF790384:0.001245,KF790408:0.00125)Node37:1e-10)Node35:1e-10,((KF886352:0.001882,(KF790438:0.000623,CY134996:0.001873)Node43:0.000623)Node41:0.001875,(KF789591:0.001879,((KF790049:0.001255,(CY134956:0.000624,CY183153:0.000624)Node51:0.000626)Node49:0.001249,((CY168407:0.004394,CY183065:0.003128)Node55:1e-10,(KF598733:0.002501,((KC892174:1e-10,(KF685747:0.000624,CY168471:0.000625)Node63:0.00375)Node61:0.000624,(((KC892685:0.000624,KC892641:1e-10)Node68:0.001874,KC893110:0.001249)Node67:1e-10,(KC892480:1e-10,(JX913043:0.001249,(KF551075:0.00125,((CY114509:1e-10,(KC892889:1e-10,KC893018:0.002498)Node81:0.005006)Node79:0.001877,(((KC892519:1e-10,KC892498:0.000624)Node86:0.002497,KC892266:0.000623)Node85:0.001026,((KC892156:0.003756,(JX978746:0.001256,KC892407:0.001883)Node93:0.001875)Node91:0.00355,((KC535402:0.000623,(KC882488:0.002498,KC535387:0.001247)Node99:0.001872)Node97:1e-10,(KC535363:0.000623,((KC535396:0.000623,(KC882883:1e-10,KC882867:0.000623)Node107:0.003124)Node105:1e-10,((KC535378:0.001872,KC535375:0.002498)Node111:1e-10,(GQ385891:0.001248,((EU779522:0.001251,(CY037727:1e-10,FJ179354:0.003124)Node119:1e-10)Node117:0.000623,((GQ385889:0.005007,(FJ686933:0.001243,FJ179356:0.003117)Node125:1e-10)Node123:0.000622,(CY173095:0.001245,(((CY173191:1e-10,CY037703:0.001247)Node132:0.000622,CY035062:0.00249)Node131:1e-10,((CY173255:0.001871,(CY044748:1e-10,GQ385846:0.001247)Node139:0.000623)Node137:1e-10,((CY027075:0.000623,(EU199367:1e-10,CY172823:0.000623)Node145:1e-10)Node143:0.000623,(((CY172847:0.000623,CY025643:1e-10)Node150:0.000623,CY172839:0.000623)Node149:0.001247,(((CY026251:0.002497,CY092241:0.00125)Node156:0.000624,CY026019:0.000623)Node155:1e-10,((CY025341:0.002502,(CY172775:0.001247,CY172903:0.000623)Node163:0.001248)Node161:0.000622,(EU199255:0.003134,(CY172431:0.000621,((CY172223:0.003127,(CY172191:0.001254,CY020069:0.000628)Node173:0.000621)Node171:1e-10,((CY092217:0.003119,(CY025485:0.001283,EU516019:0.00059)Node179:0.008843)Node177:0.001868,(CY002080:0.000622,((CY002064:0.001873,(CY002456:0.000626,CY002048:0.002491)Node187:0.001243)Node185:0.003763,(AB434109:0.006307,(((CY088198:0.001245,CY088475:0.000621)Node194:1e-10,CY000257:0.000621)Node193:0.002494,((CY112957:0.003135,CY006859:0.001253)Node199:0.000616,((CY000721:0.001243,(CY114493:1e-10,CY090885:0.002483)Node205:0.002488)Node203:1e-10,((CY001792:0.000621,(CY001600:0.00062,CY002368:0.001865)Node211:1e-10)Node209:1e-10,((CY002304:0.002489,(CY006163:0.001242,CY003632:0.00062)Node217:1e-10)Node215:0.00062,(CY001920:0.003739,(CY001912:0.002498,((CY001504:1e-10,CY006060:0.001241)Node225:0.00062,((CY001744:1e-10,CY002136:0.00124)Node229:0.00062,((CY114309:0.003746,(CY006899:0.001874,CY112901:0.001863)Node235:0.002496)Node233:0.000618,(((CY006579:0.001245,CY006283:0.000621)Node240:0.001244,CY007979:0.001864)Node239:1e-10,((CY006499:0.001241,(CY006491:0.001863,CY006635:0.00124)Node247:1e-10)Node245:0.00062,(((CY036847:0.002495,CY010004:0.001867)Node252:0.000621,CY012200:0.000621)Node251:1e-10,(((CY010028:0.001242,CY009732:0.001242)Node258:1e-10,CY010012:0.002484)Node257:0.001875,(((CY010020:0.002497,CY011416:0.001237)Node264:0.001252,CY009484:0.00062)Node263:0.005658,((CY010036:1e-10,(CY039879:0.002489,CY039880:0.000621)Node271:0.002487)Node269:0.003109,((CY010628:0.000619,(CY010716:0.002485,CY010516:0.001239)Node277:1e-10)Node275:0.000619,((CY012728:1e-10,CY013701:0.000621)Node281:0.001243,(((CY012760:0.001239,CY013200:0.001239)Node286:1e-10,CY011888:0.005614)Node285:1e-10,((CY013693:0.000619,(CY012184:0.002484,CY013669:0.00124)Node293:0.00124)Node291:1e-10,((CY112669:0.004353,(CY112556:0.001861,CY011896:0.000619)Node299:0.00186)Node297:1e-10,(CY112605:0.001243,(CY012224:0.006953,((CY012512:0.000619,(CY012896:0.00062,CY012232:0.00062)Node309:0.000621)Node307:1e-10,(CY011848:1e-10,(((CY011328:1e-10,CY114221:0.000619)Node316:0.001241,CY011560:0.00124)Node315:1e-10,((CY012456:0.000619,CY011824:0.00124)Node321:1e-10,CY017283:0.000619)Node320:1e-10)Node314:1e-10)Node312:1e-10)Node306:0.002439)Node304:0.008263)Node302:0.000618)Node296:1e-10)Node290:1e-10)Node284:0.000617)Node280:0.012609)Node274:0.000619)Node268:0.001215)Node262:0.003782)Node256:0.000613)Node250:0.006878)Node244:1e-10)Node238:0.001862)Node232:1e-10)Node228:1e-10)Node224:0.001863)Node222:0.001862)Node220:0.001245)Node214:1e-10)Node208:0.001241)Node202:0.021084)Node198:0.000623)Node192:0.00065)Node190:0.003713)Node184:0.002495)Node182:1e-10)Node176:0.005637)Node170:0.00313)Node168:0.002515)Node166:0.001876)Node160:1e-10)Node154:0.000623)Node148:1e-10)Node142:0.003749)Node136:1e-10)Node130:1e-10)Node128:0.000623)Node122:0.000623)Node116:0.001872)Node114:0.002503)Node110:1e-10)Node104:1e-10)Node102:0.000623)Node96:0.001468)Node90:0.00189)Node84:0.002721)Node78:0.002524)Node76:1e-10)Node74:0.000624)Node72:1e-10)Node66:1e-10)Node60:1e-10)Node58:0.000623)Node54:0.001248)Node48:0.001255)Node46:0.000619)Node40:1e-10)Node34:1e-10)Node28:1e-10)Node22:0.000623)Node16:1e-10)Node10:1e-10)Node4:1e-10);",
  cd2_relax = "((((Pig{Test}:0.147969,Cow{Test}:0.21343){Test}:0.085099,Horse{Test}:0.165787,Cat{Test}:0.264806):0.058611,((RhMonkey{Reference}:0.002015,Baboon{Reference}:0.003108){Reference}:0.022733,(Human{Reference}:0.004349,Chimp{Reference}:0.000799){Reference}:0.011873):0.101856):0.340802,Rat:0.050958,Mouse:0.09795);";

export {
  yokoyama, ksr2, h3, cd2_relax
};
