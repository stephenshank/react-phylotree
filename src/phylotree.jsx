import React from "react";
import { phylotree } from "phylotree";
import { scaleLinear, scaleOrdinal } from "d3-scale";
import { schemeCategory10 } from "d3-scale-chromatic";
import text_width from "./text_width";

import Branch from "./branch.jsx";

import "phylotree/build/phylotree.css";

import "./styles/phylotree.css";

function x_branch_lengths(node) {
  return node.parent ? +node.data.attribute + node.parent.data.abstract_x : 0;
}

function x_no_branch_lengths(node) {
  return node.parent ? node.parent.data.abstract_x + 1 : 0;
}

function placenodes(tree, perform_internal_layout) {
  var current_leaf_height = -1,
    unique_id = 0;
  tree.max_x = 0;
  const has_branch_lengths = Boolean(tree.get_tips()[0].data.attribute),
    x_branch_length = has_branch_lengths ? x_branch_lengths : x_no_branch_lengths;

  function node_layout(node) {
    if(!node.unique_id) {
      unique_id = node.unique_id = unique_id + 1;
    }
    node.data.abstract_x = x_branch_length(node);
    tree.max_x = Math.max(tree.max_x, node.data.abstract_x);
    if(node.children) {
      node.data.abstract_y = node.children.map(node_layout)
        .reduce( (a,b) => a + b, 0) / node.children.length;
    } else {
      current_leaf_height = node.data.abstract_y = current_leaf_height+1;
    }
    return node.data.abstract_y;
  }

  function internal_node_layout(node) {
    unique_id = node.unique_id = unique_id + 1;
    node.data.abstract_x = x_branch_length(node);
    tree.max_x = Math.max(tree.max_x, node.data.abstract_x);
    if(!tree.is_leafnode(node)) {
      node.children.forEach(internal_node_layout);
    }
    if(!node.data.abstract_y && node.data.name != "root") {
      current_leaf_height = node.data.abstract_y = current_leaf_height+1;
      tree.node_order.push(node.data.name);
    }
    if(node.parent && !node.parent.data.abstract_y && node.data.name != "root") {
      if(node.parent.data.name != "root") {
        current_leaf_height = node.parent.data.abstract_y = current_leaf_height+1;
        tree.node_order.push(node.parent.data.name);
      } 
    }
    tree.max_y = Math.max(tree.max_y, current_leaf_height);
  }

  if(perform_internal_layout) {
    tree.max_y = 0;
    tree.node_order = [];
    internal_node_layout(tree.nodes);
    const root = tree.get_node_by_name("root");
    root.data.abstract_y = root.children.map(child => child.data.abstract_y)
      .reduce((a,b)=>a+b, 0) / root.children.length;
  } else {
    node_layout(tree.nodes);
    tree.max_y = current_leaf_height;
  }
}


function Phylotree(props) {
  var{ tree, newick } = props;
  if (!tree && !newick) {
    return <g />;
  } else if(!tree) {
    tree = new phylotree(newick);
  }
  if(!props.skipPlacement) {
    placenodes(tree, props.internalNodeLabels);
  }
  const text_offset = props.showLabels ? tree.get_tips()
      .map(node => text_width(node.data.name, 14))
      .reduce((a,b) => Math.max(a,b), 0) : 0,
    padded_width = props.width - props.paddingLeft - props.paddingRight,
    padded_height = props.height - props.paddingTop - props.paddingBottom,
    x_scale = scaleLinear()
      .domain([0, tree.max_x])
      .range([0, padded_width-text_offset]),
    y_scale = scaleLinear()
      .domain([0, tree.max_y])
      .range([0, padded_height]),
    color_scale = tree.parsed_tags && props.highlightBranches ? 
      scaleOrdinal().domain(tree.parsed_tags).range(schemeCategory10) :
      null;
  return (<g transform={`translate(${props.paddingLeft}, ${props.paddingTop})`}>
    {tree.links.map(link => {
      const source_id = link.source.unique_id,
        target_id = link.target.unique_id,
        key = source_id + "," + target_id,
        show_label = props.internalNodeLabels ||
          (props.showLabels && tree.is_leafnode(link.target));
      return (<Branch
        key={key}
        xScale={x_scale}
        yScale={y_scale}
        colorScale={color_scale}
        link={link}
        maxBranchWidth={padded_width}
        showLabel={show_label}
      />);
    }) }
  </g>);
}

Phylotree.defaultProps = {
  width: 500,
  height: 500,
  paddingTop: 10,
  paddingBottom: 10,
  paddingLeft: 10,
  paddingRight: 10,
  showLabels: true,
  skipPlacement: false
};

export default Phylotree;
export {
  placenodes
};
