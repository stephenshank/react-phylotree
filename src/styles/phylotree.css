path.rp-branch {
  transition: d 0.5s;
  fill: none;
  shape-rendering: crispEdges;
}

.branch-tracer {
  transition: transform 0.5s;
}

.label {
  transition: x 0.5s y 0.5s;
}

text.label {
  font-family: Courier;
  font-size: 14px;
}

.branch-tracer {
  stroke: #bbb;
  stroke-dasharray: 3,4;
  stroke-width: 1px;
}

