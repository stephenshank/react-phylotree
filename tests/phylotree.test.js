import React from "react";
import renderer from "react-test-renderer";
import { phylotree } from "phylotree";

import Phylotree from "phylotree.jsx";
import { yokoyama } from "./data.js";


it('base phylotree renders correctly', () => {
  const tree = new phylotree(yokoyama),
    base_tree = renderer
      .create(<Phylotree tree={tree} />)
      .toJSON();
  expect(base_tree).toMatchSnapshot();
});
