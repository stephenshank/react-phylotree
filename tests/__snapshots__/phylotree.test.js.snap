// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base phylotree renders correctly 1`] = `
<g>
  <path
    className="branch"
    d="M1,738.5472222222222L1,75.7L37.3410317132809,75.7"
  />
  <path
    className="branch"
    d="M1,738.5472222222222L1,648.8322916666667L86.30183510818485,648.8322916666667"
  />
  <path
    className="branch"
    d="M1,738.5472222222222L1,1491.109375L114.1124916082064,1491.109375"
  />
  <path
    className="branch"
    d="M37.3410317132809,75.7L37.3410317132809,25.900000000000002L125.89068957446678,25.900000000000002"
  />
  <path
    className="branch"
    d="M37.3410317132809,75.7L37.3410317132809,125.5L132.0102148915072,125.5"
  />
  <path
    className="branch"
    d="M86.30183510818485,648.8322916666667L86.30183510818485,237.54999999999998L112.28825118101021,237.54999999999998"
  />
  <path
    className="branch"
    d="M86.30183510818485,648.8322916666667L86.30183510818485,1060.1145833333333L165.56274290008068,1060.1145833333333"
  />
  <path
    className="branch"
    d="M114.1124916082064,1491.109375L114.1124916082064,1395.3999999999999L328.0919063075492,1395.3999999999999"
  />
  <path
    className="branch"
    d="M114.1124916082064,1491.109375L114.1124916082064,1586.8187500000001L158.10091709215882,1586.8187500000001"
  />
  <path
    className="branch"
    d="M125.89068957446678,25.900000000000002L125.89068957446678,1L183.5072415789506,1"
  />
  <path
    className="branch"
    d="M125.89068957446678,25.900000000000002L125.89068957446678,50.800000000000004L207.56321407234498,50.800000000000004"
  />
  <path
    className="branch"
    d="M132.0102148915072,125.5L132.0102148915072,100.60000000000001L233.03241695973787,100.60000000000001"
  />
  <path
    className="branch"
    d="M132.0102148915072,125.5L132.0102148915072,150.4L209.70069628790165,150.4"
  />
  <path
    className="branch"
    d="M112.28825118101021,237.54999999999998L112.28825118101021,200.20000000000002L285.21420476592283,200.20000000000002"
  />
  <path
    className="branch"
    d="M112.28825118101021,237.54999999999998L112.28825118101021,274.90000000000003L196.85394891630384,274.90000000000003"
  />
  <path
    className="branch"
    d="M165.56274290008068,1060.1145833333333L165.56274290008068,843.96875L225.83935797389765,843.96875"
  />
  <path
    className="branch"
    d="M165.56274290008068,1060.1145833333333L165.56274290008068,1096.6000000000001L440.53919044598143,1096.6000000000001"
  />
  <path
    className="branch"
    d="M165.56274290008068,1060.1145833333333L165.56274290008068,1239.7749999999999L219.49362377664585,1239.7749999999999"
  />
  <path
    className="branch"
    d="M158.10091709215882,1586.8187500000001L158.10091709215882,1470.1000000000001L209.9782814222627,1470.1000000000001"
  />
  <path
    className="branch"
    d="M158.10091709215882,1586.8187500000001L158.10091709215882,1703.5375000000001L177.61315833235813,1703.5375000000001"
  />
  <path
    className="branch"
    d="M196.85394891630384,274.90000000000003L196.85394891630384,250L327.4013941155819,250"
  />
  <path
    className="branch"
    d="M196.85394891630384,274.90000000000003L196.85394891630384,299.8L346.44434778056313,299.8"
  />
  <path
    className="branch"
    d="M225.83935797389765,843.96875L225.83935797389765,697.1625L282.58021322854626,697.1625"
  />
  <path
    className="branch"
    d="M225.83935797389765,843.96875L225.83935797389765,990.775L249.5740371764155,990.775"
  />
  <path
    className="branch"
    d="M219.49362377664585,1239.7749999999999L219.49362377664585,1171.3L327.7146359039424,1171.3"
  />
  <path
    className="branch"
    d="M219.49362377664585,1239.7749999999999L219.49362377664585,1308.25L329.58910237064316,1308.25"
  />
  <path
    className="branch"
    d="M209.9782814222627,1470.1000000000001L209.9782814222627,1445.2L379.3826611686304,1445.2"
  />
  <path
    className="branch"
    d="M209.9782814222627,1470.1000000000001L209.9782814222627,1495L324.84906781801044,1495"
  />
  <path
    className="branch"
    d="M177.61315833235813,1703.5375000000001L177.61315833235813,1600.825L183.00191394485128,1600.825"
  />
  <path
    className="branch"
    d="M177.61315833235813,1703.5375000000001L177.61315833235813,1806.25L318.40134792222267,1806.25"
  />
  <path
    className="branch"
    d="M282.58021322854626,697.1625L282.58021322854626,571.625L310.70142768275633,571.625"
  />
  <path
    className="branch"
    d="M282.58021322854626,697.1625L282.58021322854626,822.6999999999999L351.14642524990603,822.6999999999999"
  />
  <path
    className="branch"
    d="M249.5740371764155,990.775L249.5740371764155,934.75L271.0094373032933,934.75"
  />
  <path
    className="branch"
    d="M249.5740371764155,990.775L249.5740371764155,1046.8L346.3764851164997,1046.8"
  />
  <path
    className="branch"
    d="M327.7146359039424,1171.3L327.7146359039424,1146.3999999999999L400.4404074870196,1146.3999999999999"
  />
  <path
    className="branch"
    d="M327.7146359039424,1171.3L327.7146359039424,1196.2L466.5129541577102,1196.2"
  />
  <path
    className="branch"
    d="M329.58910237064316,1308.25L329.58910237064316,1270.8999999999999L376.99404875555086,1270.8999999999999"
  />
  <path
    className="branch"
    d="M329.58910237064316,1308.25L329.58910237064316,1345.6000000000001L481.83918089935395,1345.6000000000001"
  />
  <path
    className="branch"
    d="M183.00191394485128,1600.825L183.00191394485128,1544.8L478.8620419928431,1544.8"
  />
  <path
    className="branch"
    d="M183.00191394485128,1600.825L183.00191394485128,1656.8500000000001L287.4176326211344,1656.8500000000001"
  />
  <path
    className="branch"
    d="M318.40134792222267,1806.25L318.40134792222267,1768.8999999999999L358.5756616428822,1768.8999999999999"
  />
  <path
    className="branch"
    d="M318.40134792222267,1806.25L318.40134792222267,1843.6000000000001L382.2601148058955,1843.6000000000001"
  />
  <path
    className="branch"
    d="M310.70142768275633,571.625L310.70142768275633,457.5L314.46723043095324,457.5"
  />
  <path
    className="branch"
    d="M310.70142768275633,571.625L310.70142768275633,685.75L345.88841070207746,685.75"
  />
  <path
    className="branch"
    d="M351.14642524990603,822.6999999999999L351.14642524990603,797.8000000000001L355.8511865534211,797.8000000000001"
  />
  <path
    className="branch"
    d="M351.14642524990603,822.6999999999999L351.14642524990603,847.6L366.405172732368,847.6"
  />
  <path
    className="branch"
    d="M271.0094373032933,934.75L271.0094373032933,897.4L392.7232340278761,897.4"
  />
  <path
    className="branch"
    d="M271.0094373032933,934.75L271.0094373032933,972.1L348.13746373821294,972.1"
  />
  <path
    className="branch"
    d="M376.99404875555086,1270.8999999999999L376.99404875555086,1246L499,1246"
  />
  <path
    className="branch"
    d="M376.99404875555086,1270.8999999999999L376.99404875555086,1295.8L419.1348461145226,1295.8"
  />
  <path
    className="branch"
    d="M287.4176326211344,1656.8500000000001L287.4176326211344,1619.5L318.91932791741857,1619.5"
  />
  <path
    className="branch"
    d="M287.4176326211344,1656.8500000000001L287.4176326211344,1694.2L325.4406615504955,1694.2"
  />
  <path
    className="branch"
    d="M358.5756616428822,1768.8999999999999L358.5756616428822,1744L422.82243426684977,1744"
  />
  <path
    className="branch"
    d="M358.5756616428822,1768.8999999999999L358.5756616428822,1793.8L418.9427602687838,1793.8"
  />
  <path
    className="branch"
    d="M314.46723043095324,457.5L314.46723043095324,374.5L334.7842385178707,374.5"
  />
  <path
    className="branch"
    d="M314.46723043095324,457.5L314.46723043095324,449.2L337.85607843016487,449.2"
  />
  <path
    className="branch"
    d="M314.46723043095324,457.5L314.46723043095324,548.8000000000001L332.47844155924145,548.8000000000001"
  />
  <path
    className="branch"
    d="M345.88841070207746,685.75L345.88841070207746,648.4L377.05846077567884,648.4"
  />
  <path
    className="branch"
    d="M345.88841070207746,685.75L345.88841070207746,723.1L368.1945233153281,723.1"
  />
  <path
    className="branch"
    d="M348.13746373821294,972.1L348.13746373821294,947.1999999999999L359.3394041672569,947.1999999999999"
  />
  <path
    className="branch"
    d="M348.13746373821294,972.1L348.13746373821294,997L380.49223489631123,997"
  />
  <path
    className="branch"
    d="M318.91932791741857,1619.5L318.91932791741857,1594.6000000000001L376.7785752120274,1594.6000000000001"
  />
  <path
    className="branch"
    d="M318.91932791741857,1619.5L318.91932791741857,1644.3999999999999L385.14600335055826,1644.3999999999999"
  />
  <path
    className="branch"
    d="M334.7842385178707,374.5L334.7842385178707,349.59999999999997L337.894802323218,349.59999999999997"
  />
  <path
    className="branch"
    d="M334.7842385178707,374.5L334.7842385178707,399.40000000000003L340.176828179181,399.40000000000003"
  />
  <path
    className="branch"
    d="M332.47844155924145,548.8000000000001L332.47844155924145,499L343.13939770018663,499"
  />
  <path
    className="branch"
    d="M332.47844155924145,548.8000000000001L332.47844155924145,548.8000000000001L338.3437694397054,548.8000000000001"
  />
  <path
    className="branch"
    d="M332.47844155924145,548.8000000000001L332.47844155924145,598.6L350.44977857983133,598.6"
  />
  <path
    className="branch"
    d="M368.1945233153281,723.1L368.1945233153281,698.1999999999999L377.34869827113647,698.1999999999999"
  />
  <path
    className="branch"
    d="M368.1945233153281,723.1L368.1945233153281,748L373.4287667604905,748"
  />
</g>
`;
